searx/doc.py

```diff
@@ -0,0 +1,199 @@
+
+from . import get_setting
+from .version import GIT_URL
+
+logger = logging.getLogger('doc')
+
+
+class DocPage:
+    """A page of the :py:obj:`online documentation <OnlineDoc>`."""
+
+    def __init__(self, fname, base_url=None):
+        self.fname = fname
+        self.base_url = base_url
+
+        self._raw_content = None
+        self._content = None
+        self._title = None
+
+    @property
+    def raw_content(self):
+        """Raw content of the page (without any jinja rendering)"""
+        if self._raw_content is None:
+            with open(self.fname, 'r', encoding='utf-8') as f:
+                self._raw_content = f.read()
+        return self._raw_content
+
+    @property
+    def content(self):
+        """Content of the page (rendered in a Jinja conntext)"""
+        if self._content is None:
+            ctx = self.get_ctx()
+            template = jinja2.Environment().from_string(self.raw_content)
+            self._content = template.render(**ctx)
+        return self._content
+
+    @property
+    def title(self):
+        """Title of the content (without any markup)"""
+        if self._title is None:
+            for l in self.raw_content.split('\n'):
+                if l.startswith('# '):
+                    self._title = l.strip('# ')
+                    break
+        return self._title
+
+    def get_ctx(self):  # pylint: disable=no-self-use
+        """Jinja context to render :py:obj:`DocPage.content`"""
+
+        def _md_link(name, url):
+            url = url_for(url)
+            if self.base_url:
+                url = self.base_url + url
+            return "[%s](%s)" % (name, url)
+
+        def _md_search(query):
+            url = '%s?q=%s' % (url_for('search'), urllib.parse.quote(query))
+            if self.base_url:
+                url = self.base_url + url
+            return '[%s](%s)' % (query, url)
+
+        ctx = {}
+        ctx['GIT_URL'] = GIT_URL
+        ctx['get_setting'] = get_setting
+        ctx['link'] = _md_link
+        ctx['search'] = _md_search
+
+        return ctx
+
+    def render(self):
+        """Render / return content"""
+        return self.content
+
+
+class MistletoePage(DocPage):
+    """A HTML page of the :py:obj:`online documentation <OnlineDoc>`."""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._html = None
+
+    @property
+    def html(self):
+        """HTML representation of this page"""
+        return self.render()
+
+    def render(self):
+        """Render Markdown (CommonMark_) to HTML by using mistletoe_.
+
+        .. _CommonMark: https://commonmark.org/
+        .. _mistletoe: https://github.com/miyuchina/mistletoe
+
+        """
+        if self._html is None:
+            self._html = mistletoe.markdown(self.content)
+        return self._html
+
+
+_DOC_FOLDER = os.path.join(os.path.abspath(os.path.dirname(__file__)), 'doc')
+
+
+class OnlineDoc:  # pylint: disable=too-few-public-methods
```

Comment:
With the line in mind ( from `searxng_extra/docs_prebuild` ):
```
DOC = doc.OnlineDoc(OfflinePage)  # It requires a careful reading to understand that there is no cat in a box that may be alive or dead.
```

`UserDoc`, `UserDocumentation` or a name describing the content would be better.